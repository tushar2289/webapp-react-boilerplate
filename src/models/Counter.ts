import { types } from 'mobx-state-tree';

const Counter = types
  .model({
    count: types.number,
  })
  .actions((self) => ({
    increment() {
      self.count += 1;
    },
    decrement() {
      self.count -= 1;
    },
  }));

export default Counter;
