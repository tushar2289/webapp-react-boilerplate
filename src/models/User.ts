import { types } from 'mobx-state-tree';

const TypeGeoLocation = types.model({
  lat: 'string',
  lng: 'string',
});

const TypeAddress = types.model({
  street: 'string',
  suite: 'string',
  city: 'string',
  zipcode: 'string',
  geo: TypeGeoLocation,
});

const TypeCompany = types.model({
  name: 'string',
  catchPhrase: 'string',
  bs: 'string',
});

const TypeUser = types.model({
  id: types.identifierNumber,
  name: types.string,
  username: types.string,
  email: types.string,
  address: TypeAddress,
  phone: types.string,
  website: types.string,
  company: TypeCompany,
});

export default TypeUser;
