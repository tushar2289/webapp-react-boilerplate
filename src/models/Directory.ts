import { types, flow, SnapshotIn } from 'mobx-state-tree';
import axios, { AxiosResponse } from 'axios';
import User from './User';

const TypeDirectory = types
  .model({
    users: types.map(User),
    loading: true,
  })
  .actions((self) => {
    const markLoading = (status: boolean) => {
      self.loading = status;
    };

    const processUsers = (users: SnapshotIn<typeof User>[]) => {
      users.forEach((userJson: SnapshotIn<typeof User>) => {
        return self.users.put(userJson);
      });
    };

    const loadUsers = flow(function* loadUsers() {
      markLoading(true);
      try {
        const users: AxiosResponse = yield axios.get(
          'https://jsonplaceholder.typicode.com/users'
        );
        processUsers(users.data);
        markLoading(false);
      } catch (e) {
        // Do nothing
      }
    });

    const afterCreate = () => {
      loadUsers();
    };

    return {
      afterCreate,
      loadUsers,
      processUsers,
    };
  });

export default TypeDirectory;
