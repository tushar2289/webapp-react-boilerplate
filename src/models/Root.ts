import { useContext, createContext } from 'react';
import { types, Instance } from 'mobx-state-tree';

import Counter from './Counter';
import Directory from './Directory';

const RootModel = types.model({
  counter: Counter,
  directory: Directory,
});

export const rootStore = RootModel.create({
  counter: {
    count: 0,
  },
  directory: {},
});

export type RootInstance = Instance<typeof RootModel>;
const RootStoreContext = createContext<null | RootInstance>(null);

export const { Provider } = RootStoreContext;

export function useMst(): RootInstance {
  const store = useContext(RootStoreContext);
  if (store === null) {
    throw new Error('Store cannot be null, please add a context provider');
  }
  return store;
}
