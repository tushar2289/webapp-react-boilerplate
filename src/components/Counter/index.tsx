import React from 'react';
import { observer } from 'mobx-react-lite';
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import Box from '@material-ui/core/Box';
import { useMst } from '../../models/Root';

const Counter: React.FC = observer(() => {
  const { counter } = useMst();
  return (
    <Paper elevation={3}>
      <Box p={4}>
        <Box pb={1}>
          <Typography variant="h6"> MobX powered counter</Typography>
        </Box>
        <Button variant="contained" color="primary" onClick={counter.decrement}>
          Decrement
        </Button>
        <Box px={2} component="span">
          <Typography variant="h6" component="span">
            {counter.count}
          </Typography>
        </Box>
        <Button variant="contained" color="primary" onClick={counter.increment}>
          Increment
        </Button>
      </Box>
    </Paper>
  );
});

export default Counter;
