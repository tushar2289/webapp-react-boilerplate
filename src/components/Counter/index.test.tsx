import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Counter from './index';
import { Provider, rootStore } from '../../models/Root';

afterEach(cleanup);

test('renders learn react link', () => {
  const { asFragment } = render(
    <Provider value={rootStore}>
      <Counter />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
