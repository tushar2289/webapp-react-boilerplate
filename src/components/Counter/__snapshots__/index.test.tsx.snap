// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6"
        >
           MobX powered counter
        </h6>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Decrement
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <span
        class="MuiBox-root MuiBox-root-3"
      >
        <span
          class="MuiTypography-root MuiTypography-h6"
        >
          0
        </span>
      </span>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Increment
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
